### DISCUSSION TITLE
Exposing the file descriptor in sockets

### DISCUSSION THREAD

**Discussion by ea-ae** [bluez/bluer#156] [Q&A]
I'd like to utilize the logic inside Socket to connect to a device, but afterwards handle the internal file descriptor manually (lower-level FFI). Currently my only options are either maintaining an internal fork or re-implementing the libc logic that I require.

For example, in the case of [bluer::rfcomm::Socket](https://docs.rs/bluer/latest/bluer/rfcomm/struct.Sockethtml), the internal file descriptor is private.

Would a contribution that'd introduce a `pub fn into_fd(self) -> AsyncFd<OwnedFd>` into the Socket structs be welcome? Since I don't have too much experience with open source, I thought I'd ask here beforehand. Thank you!

**Comment by surban** [comment]
Hi,

`into_raw_fd()` is [public](https://docs.rs/bluer/latest/bluer/rfcomm/struct.Socket.html#impl-IntoRawFd-for-Socket), exactly for this reason. If it does not work, you can consider it a bug and open an issue.

Hope this helps!

**Comment by surban** [comment]
However, `From<OwnedFd> for Stream` and `From<Stream> for OwnedFd` are currently not implemented. Same for `Listener`. A PR would be welcome.
